[receiver/github] add repository custom properties as resource attributes

krzko opened the pull request:
#### Description

This PR enhances the GitHub receiver by adding support for custom properties from GitHub repositories as resource attributes. The implementation extracts all key-value pairs from the repository's `CustomProperties` map and adds them as resource attributes with the prefix `github.repository.custom_properties`.

For example, if a repository has custom properties like `team_name: "open-telemetry"` and `environment: "development"`, they will be added as resource attributes:

```
github.repository.custom_properties.team_name = "open-telemetry"
github.repository.custom_properties.environment = "development"
```

The implementation handles different value types (string, int, float, bool) appropriately and follows the OpenTelemetry convention for attribute naming, similar to how HTTP headers are handled.

#### Link to tracking issue

Fixes: TBD

#### Testing

Added unit tests for the new functionality in `model_test.go`. The tests verify:

* String properties are correctly added as resource attributes
* Different types (string, int, float, bool) are handled correctly
* The "service_name" property is skipped (as it's handled separately)
* Empty and nil maps are handled properly

All tests are passing, confirming the implementation works as expected.

#### Documentation

TBD

krzko committed:
feat: add github repository custom property extraction

krzko committed:
feat: add snake_case formating for attribut keys

krzko committed:
chore: cleanup

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

krzko committed:
chore: add chloggen manifest

krzko committed:
chore: add chloggen manifest

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

adrielp reviewed (changes requested):
Overall, looks pretty good. Appreciate the addition of the tests. I left a couple nits and a suggestion around end-user documentation so that it's very clear to the end user how their custom properties are going to show up.

krzko committed:
feat: update custom property readme

krzko committed:
feat: update custom property readme

krzko commented:
> Overall, looks pretty good. Appreciate the addition of the tests. I left a couple nits and a suggestion around end-user documentation so that it's very clear to the end user how their custom properties are going to show up.

Not a problem at all.

I updated the change log gen and added a readme update about the functionality of the custom prooperties generation as well.

krzko committed:
fix: header formatting

krzko committed:
fix: update table of contents

krzko committed:
chore: fix linting and fmt

krzko committed:
fix: golden file workflow tests

krzko commented:
Fixed up the linting, formatting issues on top of the golden file tests as well.

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

adrielp commented on receiver/githubreceiver/README.md:
```diff
@@ -220,3 +221,49 @@ create a GitHub App. During the subscription phase, subscribe to `workflow_run`
+
+### Custom Properties as Resource Attributes
+
+The GitHub receiver supports adding custom properties from GitHub repositories as resource attributes in your telemetry data. This feature allows you to enrich your traces with additional metadata specific to your repositories.
```
```suggestion
The GitHub receiver supports adding custom properties from GitHub repositories as resource attributes in your telemetry data. This allows users to enrich traces and events with additional metadata specific to each repository.
```

adrielp commented on receiver/githubreceiver/README.md:
````diff
@@ -220,3 +221,49 @@ create a GitHub App. During the subscription phase, subscribe to `workflow_run`
+
+```
+team-name: platform-engineering
+environment: production
````
```suggestion
```
Let's replace the environment example with something else. Mainly because this could cause confusion with the semantic convention for deployment environments which is `deployment.environment.name` which really should come from deployment events and may not be exclusive to a single repository. Really I just want to avoid confusion on this.

krzko commented on receiver/githubreceiver/README.md:
````diff
@@ -220,3 +221,49 @@ create a GitHub App. During the subscription phase, subscribe to `workflow_run`
+
+```
+github.repository.custom_properties.team_name: "platform-engineering"
+github.repository.custom_properties.environment: "production"
````
That makes sense. I altered the examples in there.

krzko committed:
fix: github receiver readme

krzko committed:
Merge branch 'feat-receiver-github-custom-properties' of github.com:krzko/opentelemetry-collector-contrib into feat-receiver-github-custom-properties

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties

krzko committed:
Merge branch 'main' into feat-receiver-github-custom-properties
